### PR TITLE
Make diag not try ping test to cloud when there are only timestamp ch…

### PR DESCRIPTION
**Pull request by gkodali-zededa**
…anges in DevicePortConfigList & DeviceNetworkStatus

Signed-off-by: Gopi krishna Kodali <gkodali@zededa.com>

### CODE REVIEW

File: pkg/pillar/cmd/diag/diag.go [hunk 1]
```diff
@@ -343,7 +343,7 @@ func handleDNSModify(ctxArg interface{}, key string, statusArg interface{}) {
 	}
 	log.Infof("handleDNSModify for %s", key)
 	// Since we report test status we compare all fields
```
Comment (eriknordmark) [inline pkg/pillar/cmd/diag/diag.go:345]: Need to take this comment into account.
It is fine to supress the ping, but we want diag to be able to report the overall State of the DPC.

File: pkg/pillar/types/zedroutertypes.go [hunk 2]
```diff
@@ -309,6 +309,25 @@ type DevicePortConfigList struct {
...
+	if len(config.PortConfigList) != len(config2.PortConfigList) {
+		return false
+	}
+	if config.CurrentIndex != config2.CurrentIndex {
+		return false
+	}
+	for i, c1 := range config.PortConfigList {
+		c2 := config2.PortConfigList[i]
+
+		if !c1.MostlyEqual(&c2) {
```
Comment (eriknordmark) [inline pkg/pillar/types/zedroutertypes.go:324]: This does not compare the State in the DPC AFAIK.

**Review by eriknordmark — COMMENTED** [review]
I don't think we will see updates to this with this change:
// Copyright (c) 2018,2019 Zededa, Inc.
// SPDX-License-Identifier: Apache-2.0

// Utility to dump diagnostic information about connectivity

package diag

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/base64"
	"flag"
	"fmt"
	"io/ioutil"
	"mime"
	"net"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/eriknordmark/ipinfo"
	"github.com/golang/protobuf/proto"
	"github.com/google/go-cmp/cmp"
	zconfig "github.com/lf-edge/eve/api/go/config"
	"github.com/lf-edge/eve/pkg/pillar/agentlog"
	"github.com/lf-edge/eve/pkg/pillar/base"
	"github.com/lf-edge/eve/pkg/pillar/devicenetwork"
	"github.com/lf-edge/eve/pkg/pillar/hardware"
	"github.com/lf-edge/eve/pkg/pillar/pubsub"
	"github.com/lf-edge/eve/pkg/pillar/types"
	"github.com/lf-edge/eve/pkg/pillar/utils"
	"github.com/lf-edge/eve/pkg/pillar/zedcloud"
	"github.com/satori/go.uuid"
	"github.com/sirupsen/logrus"
)

const (
	agentName  = "diag"
	maxRetries = 5
	// Time limits for event loop handlers
	errorTime   = 3 * time.Minute
	warningTime = 40 * time.Second
)

// State passed to handlers
type diagContext struct {
	devicenetwork.DeviceNetworkContext
	DevicePortConfigList    *types.DevicePortConfigList
	forever                 bool // Keep on reporting until ^C
	pacContents             bool // Print PAC file contents
	ledCounter              int
	derivedLedCounter       int // Based on ledCounter + usableAddressCount
	subGlobalConfig         pubsub.Subscription
	globalConfig            *types.ConfigItemValueMap
	subLedBlinkCounter      pubsub.Subscription
	subDeviceNetworkStatus  pubsub.Subscription
	subDevicePortConfigList pubsub.Subscription
	gotBC                   bool
	gotDNS                  bool
	gotDPCList              bool
	serverNameAndPort       string
	serverName              string // Without port number
	zedcloudCtx             *zedcloud.ZedCloudContext
	cert                    *tls.Certificate
	usingOnboardCert        bool
	devUUID                 uuid.UUID
}

// Set from Makefile
var Version = "No version specified"

var debug = false
var debugOverride bool // From command line arg
var simulateDnsFailure = false
var simulatePingFailure = false
var outfile = os.Stdout
var nilUUID uuid.UUID
var logger *logrus.Logger
var log *base.LogObject

func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) int {
	logger = loggerArg
	log = logArg
	var err error
	versionPtr := flag.Bool("v", false, "Version")
	debugPtr := flag.Bool("d", false, "Debug flag")
	foreverPtr := flag.Bool("f", false, "Forever flag")
	pacContentsPtr := flag.Bool("p", false, "Print PAC file contents")
	simulateDnsFailurePtr := flag.Bool("D", false, "simulateDnsFailure flag")
	simulatePingFailurePtr := flag.Bool("P", false, "simulatePingFailure flag")
	outputFilePtr := flag.String("o", "", "file or device for output")
	flag.Parse()
	debug = *debugPtr
	debugOverride = debug
	if debugOverride {
		logger.SetLevel(logrus.TraceLevel)
	} else {
		logger.SetLevel(logrus.InfoLevel)
	}
	simulateDnsFailure = *simulateDnsFailurePtr
	simulatePingFailure = *simulatePingFailurePtr
	outputFile := *outputFilePtr
	if *versionPtr {
		fmt.Printf("%s: %s\n", os.Args[0], Version)
		return 0
	}
	if outputFile != "" {
		outfile, err = os.OpenFile(outputFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			log.Fatal(err)
		}
	}
	log.Infof("Starting %s", agentName)
	ctx := diagContext{
		forever:      *foreverPtr,
		pacContents:  *pacContentsPtr,
		globalConfig: types.DefaultConfigItemValueMap(),
	}
	ctx.AgentName = agentName
	ctx.DeviceNetworkStatus = &types.DeviceNetworkStatus{}
	ctx.DevicePortConfigList = &types.DevicePortConfigList{}

	// Make sure we have a GlobalConfig file with defaults
	utils.EnsureGCFile(log)

	// Look for global config such as log levels
	subGlobalConfig, err := ps.NewSubscription(
		pubsub.SubscriptionOptions{
			AgentName:     "",
			MyAgentName:   agentName,
			TopicImpl:     types.ConfigItemValueMap{},
			Activate:      false,
			Ctx:           &ctx,
			CreateHandler: handleGlobalConfigModify,
			ModifyHandler: handleGlobalConfigModify,
			DeleteHandler: handleGlobalConfigDelete,
			WarningTime:   warningTime,
			ErrorTime:     errorTime,
		})
	if err != nil {
		log.Fatal(err)
	}
	ctx.subGlobalConfig = subGlobalConfig
	subGlobalConfig.Activate()

	server, err := ioutil.ReadFile(types.ServerFileName)
	if err != nil {
		log.Fatal(err)
	}
	ctx.serverNameAndPort = strings.TrimSpace(string(server))
	ctx.serverName = strings.Split(ctx.serverNameAndPort, ":")[0]

	zedcloudCtx := zedcloud.NewContext(log, zedcloud.ContextOptions{
		DevNetworkStatus: ctx.DeviceNetworkStatus,
		Timeout:          ctx.globalConfig.GlobalValueInt(types.NetworkSendTimeout),
		NeedStatsFunc:    true,
		Serial:           hardware.GetProductSerial(log),
		SoftSerial:       hardware.GetSoftSerial(log),
		AgentName:        agentName,
	})
	log.Infof("Diag Get Device Serial %s, Soft Serial %s", zedcloudCtx.DevSerial,
		zedcloudCtx.DevSoftSerial)

	// XXX move to later for Get UUID if available

	log.Infof("diag Run: Use V2 API %v", zedcloudCtx.V2API)

	if fileExists(types.DeviceCertName) {
		// Load device cert
		cert, err := zedcloud.GetClientCert()
		if err != nil {
			log.Fatal(err)
		}
		ctx.cert = &cert
	} else if fileExists(types.OnboardCertName) &&
		fileExists(types.OnboardKeyName) {
		cert, err := tls.LoadX509KeyPair(types.OnboardCertName,
			types.OnboardKeyName)
		if err != nil {
			log.Fatal(err)
		}
		ctx.cert = &cert
		fmt.Fprintf(outfile, "WARNING: no device cert; using onboarding cert at %v\n",
			time.Now().Format(time.RFC3339Nano))
		ctx.usingOnboardCert = true
	} else {
		fmt.Fprintf(outfile, "ERROR: no device cert and no onboarding cert at %v\n",
			time.Now().Format(time.RFC3339Nano))
		return 1
	}
	ctx.zedcloudCtx = &zedcloudCtx

	subLedBlinkCounter, err := ps.NewSubscription(
		pubsub.SubscriptionOptions{
			AgentName:     "",
			MyAgentName:   agentName,
			TopicImpl:     types.LedBlinkCounter{},
			Activate:      false,
			Ctx:           &ctx,
			CreateHandler: handleLedBlinkModify,
			ModifyHandler: handleLedBlinkModify,
			WarningTime:   warningTime,
			ErrorTime:     errorTime,
		})
	if err != nil {
		errStr := fmt.Sprintf("ERROR: internal Subscribe failed %s\n", err)
		panic(errStr)
	}
	ctx.subLedBlinkCounter = subLedBlinkCounter
	subLedBlinkCounter.Activate()

	subDeviceNetworkStatus, err := ps.NewSubscription(
		pubsub.SubscriptionOptions{
			AgentName:     "nim",
			MyAgentName:   agentName,
			TopicImpl:     types.DeviceNetworkStatus{},
			Activate:      false,
			Ctx:           &ctx,
			CreateHandler: handleDNSModify,
			ModifyHandler: handleDNSModify,
			DeleteHandler: handleDNSDelete,
			WarningTime:   warningTime,
			ErrorTime:     errorTime,
		})
	if err != nil {
		errStr := fmt.Sprintf("ERROR: internal Subscribe failed %s\n", err)
		panic(errStr)
	}
	ctx.subDeviceNetworkStatus = subDeviceNetworkStatus
	subDeviceNetworkStatus.Activate()

	subDevicePortConfigList, err := ps.NewSubscription(
		pubsub.SubscriptionOptions{
			AgentName:     "nim",
			MyAgentName:   agentName,
			Persistent:    true,
			TopicImpl:     types.DevicePortConfigList{},
			Activate:      false,
			Ctx:           &ctx,
			CreateHandler: handleDPCModify,
			ModifyHandler: handleDPCModify,
		})
	if err != nil {
		errStr := fmt.Sprintf("ERROR: internal Subscribe failed %s\n", err)
		panic(errStr)
	}
	ctx.subDevicePortConfigList = subDevicePortConfigList
	subDevicePortConfigList.Activate()

	subOnboardStatus, err := ps.NewSubscription(pubsub.SubscriptionOptions{
		AgentName:     "zedclient",
		MyAgentName:   agentName,
		CreateHandler: handleOnboardStatusModify,
		ModifyHandler: handleOnboardStatusModify,
		WarningTime:   warningTime,
		ErrorTime:     errorTime,
		TopicImpl:     types.OnboardingStatus{},
		Activate:      true,
		Persistent:    true,
		Ctx:           &ctx,
	})
	if err != nil {
		log.Fatal(err)
	}

	for {
		select {
		case change := <-subGlobalConfig.MsgChan():
			subGlobalConfig.ProcessChange(change)

		case change := <-subLedBlinkCounter.MsgChan():
			ctx.gotBC = true
			subLedBlinkCounter.ProcessChange(change)

		case change := <-subDeviceNetworkStatus.MsgChan():
			ctx.gotDNS = true
			subDeviceNetworkStatus.ProcessChange(change)

		case change := <-subOnboardStatus.MsgChan():
			subOnboardStatus.ProcessChange(change)

		case change := <-subDevicePortConfigList.MsgChan():
			ctx.gotDPCList = true
			subDevicePortConfigList.ProcessChange(change)
		}
		if !ctx.forever && ctx.gotDNS && ctx.gotBC && ctx.gotDPCList {
			break
		}
		if ctx.usingOnboardCert && fileExists(types.DeviceCertName) {
			fmt.Fprintf(outfile, "WARNING: Switching from onboard to device cert\n")
			// Load device cert
			cert, err := zedcloud.GetClientCert()
			if err != nil {
				log.Fatal(err)
			}
			ctx.cert = &cert
			ctx.usingOnboardCert = false
		}
	}
	return 0
}

func fileExists(filename string) bool {
	_, err := os.Stat(filename)
	return err == nil
}

// Handles both create and modify events
func handleLedBlinkModify(ctxArg interface{}, key string,
	configArg interface{}) {

	config := configArg.(types.LedBlinkCounter)
	ctx := ctxArg.(*diagContext)

	if key != "ledconfig" {
		log.Errorf("handleLedBlinkModify: ignoring %s", key)
		return
	}
	// Supress work and logging if no change
	if config.BlinkCounter == ctx.ledCounter {
		return
	}
	ctx.ledCounter = config.BlinkCounter
	ctx.derivedLedCounter = types.DeriveLedCounter(ctx.ledCounter,
		ctx.UsableAddressCount)
	log.Infof("counter %d usableAddr %d, derived %d",
		ctx.ledCounter, ctx.UsableAddressCount, ctx.derivedLedCounter)
	// XXX wait in case we get another handle call?
	// XXX set output sched in ctx; print one second later?
	printOutput(ctx)
}

// Handles both create and modify events
func handleDNSModify(ctxArg interface{}, key string, statusArg interface{}) {

	status := statusArg.(types.DeviceNetworkStatus)
	ctx := ctxArg.(*diagContext)
	if key != "global" {
		log.Infof("handleDNSModify: ignoring %s", key)
		return
	}
	log.Infof("handleDNSModify for %s", key)
	// Since we report test status we compare all fields
	if cmp.Equal(ctx.DeviceNetworkStatus, status) {
		log.Infof("handleDNSModify unchanged")
		return
	}
	log.Infof("handleDNSModify: changed %v",
		cmp.Diff(ctx.DeviceNetworkStatus, status))
	*ctx.DeviceNetworkStatus = status
	newAddrCount := types.CountLocalAddrAnyNoLinkLocal(*ctx.DeviceNetworkStatus)
	log.Infof("handleDNSModify %d usable addresses", newAddrCount)
	if (ctx.UsableAddressCount == 0 && newAddrCount != 0) ||
		(ctx.UsableAddressCount != 0 && newAddrCount == 0) {
		ctx.UsableAddressCount = newAddrCount
		ctx.derivedLedCounter = types.DeriveLedCounter(ctx.ledCounter,
			ctx.UsableAddressCount)
		log.Infof("counter %d usableAddr %d, derived %d",
			ctx.ledCounter, ctx.UsableAddressCount, ctx.derivedLedCounter)
	}

	// update proxy certs if configured
	if ctx.zedcloudCtx != nil && ctx.zedcloudCtx.V2API {
		zedcloud.UpdateTLSProxyCerts(ctx.zedcloudCtx)
	}
	// XXX can we limit to interfaces which changed?
	// XXX wait in case we get another handle call?
	// XXX set output sched in ctx; print one second later?
	printOutput(ctx)
	log.Infof("handleDNSModify done for %s", key)
}

func handleDNSDelete(ctxArg interface{}, key string,
	statusArg interface{}) {

	log.Infof("handleDNSDelete for %s", key)
	ctx := ctxArg.(*diagContext)

	if key != "global" {
		log.Infof("handleDNSDelete: ignoring %s", key)
		return
	}
	*ctx.DeviceNetworkStatus = types.DeviceNetworkStatus{}
	newAddrCount := types.CountLocalAddrAnyNoLinkLocal(*ctx.DeviceNetworkStatus)
	log.Infof("handleDNSDelete %d usable addresses", newAddrCount)
	if (ctx.UsableAddressCount == 0 && newAddrCount != 0) ||
		(ctx.UsableAddressCount != 0 && newAddrCount == 0) {
		ctx.UsableAddressCount = newAddrCount
		ctx.derivedLedCounter = types.DeriveLedCounter(ctx.ledCounter,
			ctx.UsableAddressCount)
		log.Infof("counter %d usableAddr %d, derived %d",
			ctx.ledCounter, ctx.UsableAddressCount, ctx.derivedLedCounter)
	}
	// XXX wait in case we get another handle call?
	// XXX set output sched in ctx; print one second later?
	printOutput(ctx)
	log.Infof("handleDNSDelete done for %s", key)
}

// Handles both create and modify events
func handleDPCModify(ctxArg interface{}, key string, statusArg interface{}) {

	status := statusArg.(types.DevicePortConfigList)
	ctx := ctxArg.(*diagContext)
	if key != "global" {
		log.Infof("handleDPCModify: ignoring %s", key)
		return
	}
	log.Infof("handleDPCModify for %s", key)
	if cmp.Equal(ctx.DevicePortConfigList, status) {
		return
	}
	log.Infof("handleDPCModify: changed %v",
		cmp.Diff(ctx.DevicePortConfigList, status))
	*ctx.DevicePortConfigList = status
	// XXX can we limit to interfaces which changed?
	// XXX exclude if only timestamps changed?
	// XXX wait in case we get another handle call?
	// XXX set output sched in ctx; print one second later?
	printOutput(ctx)
	log.Infof("handleDPCModify done for %s", key)
}

// Handles UUID change from process client
func handleOnboardStatusModify(ctxArg interface{}, key string, statusArg interface{}) {
	status := statusArg.(types.OnboardingStatus)
	ctx := ctxArg.(*diagContext)
	if cmp.Equal(ctx.devUUID, status.DeviceUUID) {
		log.Infof("handleOnboardStatusModify no change to %v", ctx.devUUID)
		return
	}
	ctx.devUUID = status.DeviceUUID
	log.Infof("handleOnboardStatusModify changed to %v", ctx.devUUID)
	printOutput(ctx)
}

// Print output for all interfaces
// XXX can we limit to interfaces which changed?
func printOutput(ctx *diagContext) {

	// Defer until we have an initial BlinkCounter and DeviceNetworkStatus
	if !ctx.gotDNS || !ctx.gotBC || !ctx.gotDPCList {
		return
	}

	fmt.Fprintf(outfile, "\nINFO: updated diag information at %v\n",
		time.Now().Format(time.RFC3339Nano))
	// XXX certificate fingerprints? What does zedcloud use?
	if fileExists(types.SelfRegFile) {
		fmt.Fprintf(outfile, "INFO: selfRegister is still in progress\n")
		// XXX print onboarding cert
	}

	switch ctx.derivedLedCounter {
	case 0:
		fmt.Fprintf(outfile, "ERROR: Summary: Unknown LED counter 0\n")
	case 1:
		fmt.Fprintf(outfile, "ERROR: Summary: Waiting for DHCP IP address(es)\n")
	case 2:
		fmt.Fprintf(outfile, "ERROR: Summary: Trying to connect to EV Controller\n")
	case 3:
		fmt.Fprintf(outfile, "WARNING: Summary: Connected to EV Controller but not onboarded\n")
	case 4:
		fmt.Fprintf(outfile, "INFO: Summary: Connected to EV Controller and onboarded\n")
	case 10:
		fmt.Fprintf(outfile, "ERROR: Summary: Onboarding failure or conflict\n")
	case 12:
		fmt.Fprintf(outfile, "ERROR: Summary: Response without TLS - ignored\n")
	case 13:
		fmt.Fprintf(outfile, "ERROR: Summary: Response without OSCP or bad OSCP - ignored\n")
	default:
		fmt.Fprintf(outfile, "ERROR: Summary: Unsupported LED counter %d\n",
			ctx.derivedLedCounter)
	}

	testing := ctx.DeviceNetworkStatus.Testing
	var upcase, downcase string
	if testing {
		upcase = "Testing"
		downcase = "testing"
	} else {
		upcase = "Using"
		downcase = "using"
	}
	// Print info about fallback
	DPCLen := len(ctx.DevicePortConfigList.PortConfigList)
	if DPCLen > 0 {
		first := ctx.DevicePortConfigList.PortConfigList[0]
		if ctx.DevicePortConfigList.CurrentIndex == -1 {
			fmt.Fprintf(outfile, "WARNING: Have no currently working DevicePortConfig\n")
		} else if ctx.DevicePortConfigList.CurrentIndex != 0 {
			fmt.Fprintf(outfile, "WARNING: Not %s highest priority DevicePortConfig key %s due to %s\n",
				downcase, first.Key, first.LastError)
			for i, dpc := range ctx.DevicePortConfigList.PortConfigList {
				if i == 0 {
					continue
				}
				if i != ctx.DevicePortConfigList.CurrentIndex {
					fmt.Fprintf(outfile, "WARNING: Not %s priority %d DevicePortConfig key %s due to %s\n",
						downcase, i, dpc.Key, dpc.LastError)
				} else {
					fmt.Fprintf(outfile, "INFO: %s priority %d DevicePortConfig key %s\n",
						upcase, i, dpc.Key)
					break
				}
			}
			if DPCLen-1 > ctx.DevicePortConfigList.CurrentIndex {
				fmt.Fprintf(outfile, "INFO: Have %d backup DevicePortConfig\n",
					DPCLen-1-ctx.DevicePortConfigList.CurrentIndex)
			}
		} else {
			fmt.Fprintf(outfile, "INFO: %s highest priority DevicePortConfig key %s\n",
				upcase, first.Key)
			if DPCLen > 1 {
				fmt.Fprintf(outfile, "INFO: Have %d backup DevicePortConfig\n",
					DPCLen-1)
			}
		}
	}
	if testing {
		fmt.Fprintf(outfile, "WARNING: The configuration below is under test hence might report failures\n")
	}
	if ctx.DeviceNetworkStatus.State != types.DPC_SUCCESS {
		fmt.Fprintf(outfile, "WARNING: state %s not SUCCESS\n",
			ctx.DeviceNetworkStatus.State.String())
	}
can you check?

File: pkg/pillar/cmd/diag/diag.go (hunk 1 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/diag/diag.go:345]: So, should I revert this back to original code?

**Comment by rvs** [comment]
Hey @eriknordmark  It appears that the New Yetus's golangcilint is coredumping on this PR -- this is kind of fun (and @aw-was-here and I are looking into this) but it shouldn't prevent you from merging it.

File: pkg/pillar/cmd/diag/diag.go (hunk 1 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/diag/diag.go:345]: Will revert back to original code.

**Comment by gkodali-zededa** [comment]
> I don't think we will see diag printing these warnings and not printing them when the state changes:
> if testing {
> fmt.Fprintf(outfile, "WARNING: The configuration below is under test hence might report failures\n")
> }
> if ctx.DeviceNetworkStatus.State != types.DPC_SUCCESS {
> fmt.Fprintf(outfile, "WARNING: state %s not SUCCESS\n",
> ctx.DeviceNetworkStatus.State.String())
> }
> 
> can you check?

I did not understand what I have to check here. Can you please elaborate?

**Comment by eriknordmark** [comment]
> I did not understand what I have to check here. Can you please elaborate?

Verify (by reading the code or testing) that a change to ctx.DeviceNetworkStatus.State or the testing field will still result in something printed by diag.

**Review by eriknordmark — COMMENTED** [review]
Does this more limited change reduce the times diag prints things?
It doesn't touch the State and Testing parts in DeviceNetworkStatus so it shouldn't have the concerns I raised.

**Comment by gkodali-zededa** [comment]
> > I did not understand what I have to check here. Can you please elaborate?
> 
> Verify (by reading the code or testing) that a change to ctx.DeviceNetworkStatus.State or the testing field will still result in something printed by diag.

I have reverted handleDNSModify in diag to use old cmp module for comparing old and new device network status.
But, if we use DeviceNetworkStatus.MostlyEqual for comparison between old and new DNS state, we will not see diag printing anything when - only the "State" or "Testing" fields of DNS change. That is because DeviceNetworkStatus.MostlyEqual does not compare "State", "Testing" fields between old and new device network statuses.

**Comment by gkodali-zededa** [comment]
> Does this more limited change reduce the times diag prints things?
> It doesn't touch the State and Testing parts in DeviceNetworkStatus so it shouldn't have the concerns I raised.

When using DeviceNetworkStatus.MosltyEqual for comparison, there was very little printing from diag.

**Pull request merged**